### ISSUE TITLE
corrupted upgrade from 21.1: /usr/local/libexec/opnsense-auth: not found

### ISSUE THREAD

**Issue by comotion** [opnsense/core#5188]
**Important notices**

Before you add a new report, we ask you kindly to acknowledge the following:

- [x] I have read the contributing guide lines at https://github.com/opnsense/core/blob/master/CONTRIBUTING.md
- [x] I am convinced that my issue is new after having checked both open and closed issues at https://github.com/opnsense/core/issues?q=is%3Aissue

**Describe the bug**

Tried upgrading our gateway to 21.7 yesterday, from 20.7. Made it as far as 21.1, and then had an upgrade failure on the console which resulted in a broken system. Please see [opnsense-21.7-upgrade-fail.log](https://github.com/opnsense/core/files/7082199/opnsense-21.7-upgrade-fail.log)

The GUI is still accessible, but there is no way to get administrative access because of the message `sh: /usr/local/libexec/opnsense-auth: not found`.

Hardware is HP DL20 Gen10.

**Comment by AdSchellevis** [comment]
looks like using third party (FreeBSD) package mirrors, which aren't compatible.
```
	varnish6: 6.5.1 -> 6.6.1 [FreeBSD]
```

reinstall is likely the quickest fix...

https://docs.opnsense.org/manual/software_included.html#packages-pkg

**Comment by comotion** [comment]
thanks, that explains it at least.

**Comment by akama-aka** [comment]
Same with 24.x

**Comment by OdinVex** [comment]
Happened again, June 2nd.

**Comment by fichtner** [comment]
More details please. Versions first and foremost

**Comment by OdinVex** [comment]
Able to access SSH (no root/sudo), 26.1.6? Community.

Version: `26.1` Edit: Assuming `.8_5`, not sure, can't recall how to fetch.
Mirror: `https://pkg.opnsense.org`
Plugins: `os-acme-client,os-cpu-microcode-intel,os-ddclient,os-dmidecode,os-freeradius,os-git-backup,os-iperf,os-isc-dhcp,os-smart,os-theme-rebellion,os-turnserver,os-vnstat`

All Web UI endpoints 403 (no forms, nothing), no sudo in SSH.

**Comment by fichtner** [comment]
As far as I know the bug that could cause this was fixed in earlier 26.1.x, maybe .4

**Comment by OdinVex** [comment]
Was fine last week, so I'm baffled. No config change (checked /conf). Edit: There's also no `opnsense-shell`.
Edit: I misread the version, it's only showing `26.1` and I can't recall how to fetch the full version currently installed.

**Comment by OdinVex** [comment]
Just going to reinstall again, second time it's happened (first incident was before fix was released which doesn't seem to have worked).

**Comment by fichtner** [comment]
Without details this is meaningless.


Cheers,
Franco

**Comment by OdinVex** [comment]
> Without details this is meaningless.
> 
> Cheers, Franco

Gave what I could above. Unable to actually cat much or interact with anything with sudo being borked. Was hoping to try to find update logs to see if anything was missed, unable to navigate the filesystem. Not meaningless, just probably not as hoped helpful.

**Comment by fichtner** [comment]
# opnsense-update -g

This may yield the relevant update log. There’s a file as well, but opnsense-update should still be installed if core is missing or not fully installed.

Apart from the bug that would prematurely reboot an unfinished update there can be other reasons why packages fail to install on the disk.


Cheers,
Franco

**Comment by OdinVex** [comment]
> ```
> # opnsense-update -g
> ```
> 
> This may yield the relevant update log. There’s a file as well, but opnsense-update should still be installed if core is missing or not fully installed.
> 
> Apart from the bug that would prematurely reboot an unfinished update there can be other reasons why packages fail to install on the disk.
> 
> Cheers, Franco

Unfortunately `opnsense-update -g` requires root, too. I'm concerned about rebooting it (it hasn't rebooted since the update) because I'm not sure how much useful state info (eg will it be able to boot again in an interactive way, will it even start SSH, etc) may be lost, otherwise I'd try single-user and go from there to try to find out what happened to help report.

Edit: Managed to find the update logs on each firewall, both were running `OPNsense 26.1.8_5` when it happened. One firewall has an extra repo, but the primary one doesn't, both have the issue, so I'll post the one with the extra repo:

```
***GOT REQUEST TO UPDATE***
Currently running OPNsense 26.1.8_5 (amd64) at Wed Jun  3 03:21:23 EDT 2026
Updating OPNsense repository catalogue...
Fetching meta.conf: . done
Fetching data.pkg: .......... done
Processing entries: .......... done
OPNsense repository update completed. 927 packages processed.
Updating mimugmail repository catalogue...
Fetching meta.conf: . done
mimugmail repository is up to date.
All repositories are up to date.
Updating OPNsense repository catalogue...
OPNsense repository is up to date.
Updating mimugmail repository catalogue...
Fetching meta.conf: . done
mimugmail repository is up to date.
All repositories are up to date.
Checking for upgrades (110 candidates): .......... done
Processing candidates (110 candidates): .... done
The following 42 package(s) will be affected (of 0 checked):

New packages to be INSTALLED:
        py313-numpy: 2.4.4_3,1 [OPNsense]

Installed packages to be UPGRADED:
        ca_root_nss: 3.117_2 -> 3.124 [OPNsense]
        cpu-microcode-intel: 20260227 -> 20260512 [OPNsense]
        gettext-runtime: 1.0 -> 1.0_1 [OPNsense]
        libgpg-error: 1.60 -> 1.61 [OPNsense]
        nspr: 4.38.2 -> 4.39 [OPNsense]
        nss: 3.123.1 -> 3.124 [OPNsense]
        openvpn: 2.6.20 -> 2.7.4 [OPNsense]
        opnsense: 26.1.8_5 -> 26.1.9 [OPNsense]
        opnsense-update: 26.1.7_1 -> 26.1.9 [OPNsense]
        os-isc-dhcp: 1.0_4 -> 1.0_5 [OPNsense]
        pciids: 20260430 -> 20260514 [OPNsense]
        php83: 8.3.30 -> 8.3.31 [OPNsense]
        php83-ctype: 8.3.30 -> 8.3.31 [OPNsense]
        php83-curl: 8.3.30 -> 8.3.31 [OPNsense]
        php83-dom: 8.3.30 -> 8.3.31 [OPNsense]
        php83-filter: 8.3.30 -> 8.3.31 [OPNsense]
        php83-gettext: 8.3.30 -> 8.3.31 [OPNsense]
        php83-ldap: 8.3.30 -> 8.3.31 [OPNsense]
        php83-mbstring: 8.3.30 -> 8.3.31 [OPNsense]
        php83-pcntl: 8.3.30 -> 8.3.31 [OPNsense]
        php83-pdo: 8.3.30 -> 8.3.31 [OPNsense]
        php83-session: 8.3.30 -> 8.3.31 [OPNsense]
        php83-simplexml: 8.3.30 -> 8.3.31 [OPNsense]
        php83-sockets: 8.3.30 -> 8.3.31 [OPNsense]
        php83-sqlite3: 8.3.30 -> 8.3.31 [OPNsense]
        php83-xml: 8.3.30 -> 8.3.31 [OPNsense]
        php83-zlib: 8.3.30 -> 8.3.31 [OPNsense]
        py313-boto3: 1.43.2 -> 1.43.7 [OPNsense]
        py313-botocore: 1.43.2 -> 1.43.7 [OPNsense]
        py313-bottleneck: 1.6.0_1 -> 1.6.0_2 [OPNsense]
        py313-duckdb: 1.5.2 -> 1.5.2_1 [OPNsense]
        py313-idna: 3.13 -> 3.15 [OPNsense]
        py313-numexpr: 2.14.1_1 -> 2.14.1_2 [OPNsense]
        py313-pandas: 2.3.3_1,1 -> 2.3.3_2,1 [OPNsense]
        py313-pytz: 2025.2_1,1 -> 2026.2,1 [OPNsense]
        py313-requests: 2.33.1 -> 2.34.2 [OPNsense]
        py313-ujson: 5.12.0 -> 5.12.1 [OPNsense]
        py313-urllib3: 2.6.3,1 -> 2.7.0,1 [OPNsense]
        python313: 3.13.13 -> 3.13.13_1 [OPNsense]
        suricata: 8.0.4 -> 8.0.5 [OPNsense]
        unbound: 1.25.0 -> 1.25.1 [OPNsense]

Number of packages to be installed: 1
Number of packages to be upgraded: 41

The process will require 49 MiB more space.
137 MiB to be downloaded.
[1/42] Fetching unbound-1.25.1.pkg: .......... done
[2/42] Fetching py313-idna-3.15.pkg: .......... done
[3/42] Fetching php83-filter-8.3.31.pkg: ... done
[4/42] Fetching pciids-20260514.pkg: .......... done
[5/42] Fetching opnsense-update-26.1.9.pkg: ..... done
[6/42] Fetching py313-boto3-1.43.7.pkg: .......... done
[7/42] Fetching php83-curl-8.3.31.pkg: ....... done
[8/42] Fetching nss-3.124.pkg: .......... done
[9/42] Fetching php83-ldap-8.3.31.pkg: ..... done
[10/42] Fetching os-isc-dhcp-1.0_5.pkg: ..... done
[11/42] Fetching python313-3.13.13_1.pkg: .......... done
[12/42] Fetching openvpn-2.7.4.pkg: .......... done
[13/42] Fetching py313-pandas-2.3.3_2,1.pkg: .......... done
[14/42] Fetching php83-simplexml-8.3.31.pkg: .... done
[15/42] Fetching php83-pdo-8.3.31.pkg: ........ done
[16/42] Fetching py313-pytz-2026.2,1.pkg: .......... done
[17/42] Fetching php83-sockets-8.3.31.pkg: ...... done
[18/42] Fetching py313-bottleneck-1.6.0_2.pkg: .......... done
[19/42] Fetching php83-pcntl-8.3.31.pkg: ... done
[20/42] Fetching ca_root_nss-3.124.pkg: .......... done
[21/42] Fetching php83-sqlite3-8.3.31.pkg: .... done
[22/42] Fetching py313-botocore-1.43.7.pkg: .......... done
[23/42] Fetching php83-session-8.3.31.pkg: ...... done
[24/42] Fetching php83-mbstring-8.3.31.pkg: .......... done
[25/42] Fetching php83-gettext-8.3.31.pkg: . done
[26/42] Fetching php83-zlib-8.3.31.pkg: ... done
[27/42] Fetching php83-ctype-8.3.31.pkg: . done
[28/42] Fetching libgpg-error-1.61.pkg: .......... done
[29/42] Fetching nspr-4.39.pkg: .......... done
[30/42] Fetching gettext-runtime-1.0_1.pkg: .......... done
[31/42] Fetching php83-8.3.31.pkg: .......... done
[32/42] Fetching py313-urllib3-2.7.0,1.pkg: .......... done
[33/42] Fetching php83-xml-8.3.31.pkg: .... done
[34/42] Fetching py313-requests-2.34.2.pkg: .......... done
[35/42] Fetching php83-dom-8.3.31.pkg: .......... done
[36/42] Fetching suricata-8.0.5.pkg: .......... done
[37/42] Fetching py313-numexpr-2.14.1_2.pkg: .......... done
[38/42] Fetching py313-ujson-5.12.1.pkg: ....... done
[39/42] Fetching opnsense-26.1.9.pkg: .......... done
[40/42] Fetching cpu-microcode-intel-20260512.pkg: .......... done
[41/42] Fetching py313-numpy-2.4.4_3,1.pkg: .......... done
[42/42] Fetching py313-duckdb-1.5.2_1.pkg: .......... done
Checking integrity... done (1 conflicting)
  - py313-numpy-2.4.4_3,1 [OPNsense] conflicts with py313-numpy1-1.26.4_3 [installed] on /usr/local/bin/f2py
Checking integrity... done (0 conflicting)
Conflicts with the existing packages have been found.
One more solver iteration is needed to resolve them.
The following 43 package(s) will be affected (of 0 checked):

New packages to be INSTALLED:
        py313-numpy: 2.4.4_3,1 [OPNsense]

Installed packages to be UPGRADED:
        ca_root_nss: 3.117_2 -> 3.124 [OPNsense]
        cpu-microcode-intel: 20260227 -> 20260512 [OPNsense]
        gettext-runtime: 1.0 -> 1.0_1 [OPNsense]
        libgpg-error: 1.60 -> 1.61 [OPNsense]
        nspr: 4.38.2 -> 4.39 [OPNsense]
        nss: 3.123.1 -> 3.124 [OPNsense]
        openvpn: 2.6.20 -> 2.7.4 [OPNsense]
        opnsense: 26.1.8_5 -> 26.1.9 [OPNsense]
        opnsense-update: 26.1.7_1 -> 26.1.9 [OPNsense]
        os-isc-dhcp: 1.0_4 -> 1.0_5 [OPNsense]
        pciids: 20260430 -> 20260514 [OPNsense]
        php83: 8.3.30 -> 8.3.31 [OPNsense]
        php83-ctype: 8.3.30 -> 8.3.31 [OPNsense]
        php83-curl: 8.3.30 -> 8.3.31 [OPNsense]
        php83-dom: 8.3.30 -> 8.3.31 [OPNsense]
        php83-filter: 8.3.30 -> 8.3.31 [OPNsense]
        php83-gettext: 8.3.30 -> 8.3.31 [OPNsense]
        php83-ldap: 8.3.30 -> 8.3.31 [OPNsense]
        php83-mbstring: 8.3.30 -> 8.3.31 [OPNsense]
        php83-pcntl: 8.3.30 -> 8.3.31 [OPNsense]
        php83-pdo: 8.3.30 -> 8.3.31 [OPNsense]
        php83-session: 8.3.30 -> 8.3.31 [OPNsense]
        php83-simplexml: 8.3.30 -> 8.3.31 [OPNsense]
        php83-sockets: 8.3.30 -> 8.3.31 [OPNsense]
        php83-sqlite3: 8.3.30 -> 8.3.31 [OPNsense]
        php83-xml: 8.3.30 -> 8.3.31 [OPNsense]
        php83-zlib: 8.3.30 -> 8.3.31 [OPNsense]
        py313-boto3: 1.43.2 -> 1.43.7 [OPNsense]
        py313-botocore: 1.43.2 -> 1.43.7 [OPNsense]
        py313-bottleneck: 1.6.0_1 -> 1.6.0_2 [OPNsense]
        py313-duckdb: 1.5.2 -> 1.5.2_1 [OPNsense]
        py313-idna: 3.13 -> 3.15 [OPNsense]
        py313-numexpr: 2.14.1_1 -> 2.14.1_2 [OPNsense]
        py313-pandas: 2.3.3_1,1 -> 2.3.3_2,1 [OPNsense]
        py313-pytz: 2025.2_1,1 -> 2026.2,1 [OPNsense]
        py313-requests: 2.33.1 -> 2.34.2 [OPNsense]
        py313-ujson: 5.12.0 -> 5.12.1 [OPNsense]
        py313-urllib3: 2.6.3,1 -> 2.7.0,1 [OPNsense]
        python313: 3.13.13 -> 3.13.13_1 [OPNsense]
        suricata: 8.0.4 -> 8.0.5 [OPNsense]
        unbound: 1.25.0 -> 1.25.1 [OPNsense]

Installed packages to be REMOVED:
        py313-numpy1: 1.26.4_3

Number of packages to be removed: 1
Number of packages to be installed: 1
Number of packages to be upgraded: 41

The process will require 5 MiB more space.
Checking integrity... done (0 conflicting)
[1/48] Upgrading cpu-microcode-intel from 20260227 to 20260512...
[1/48] Extracting cpu-microcode-intel-20260512: .......... done
[2/48] Upgrading gettext-runtime from 1.0 to 1.0_1...
[2/48] Extracting gettext-runtime-1.0_1: .......... done
[3/48] Upgrading libgpg-error from 1.60 to 1.61...
[3/48] Extracting libgpg-error-1.61: .......... done
[4/48] Upgrading nspr from 4.38.2 to 4.39...
[4/48] Extracting nspr-4.39: .......... done
[5/48] Upgrading nss from 3.123.1 to 3.124...
[5/48] Extracting nss-3.124: .......... done
[6/48] Upgrading os-isc-dhcp from 1.0_4 to 1.0_5...
[6/48] Extracting os-isc-dhcp-1.0_5: .......... done
Stopping configd...done
Starting configd.

Fatal error: Uncaught Error: Class "OPNsense\Base\Menu\MenuContainer" not found in /usr/local/opnsense/mvc/app/models/OPNsense/DHCPv4/Menu/Menu.php:34
Stack trace:
#0 /usr/local/opnsense/mvc/app/library/OPNsense/Autoload/Loader.php(49): require_once()
#1 [internal function]: OPNsense\Autoload\Loader->autoload('OPNsense\\DHCPv4...')
#2 /usr/local/opnsense/mvc/script/run_migrations.php(64): ReflectionClass->__construct('OPNsense\\DHCPv4...')
#3 {main}
  thrown in /usr/local/opnsense/mvc/app/models/OPNsense/DHCPv4/Menu/Menu.php on line 34
Reloading plugin configuration
Flushing all caches...done.
Configuring system logging...done.
Reloading template OPNsense/Syslog: OK
[7/48] Upgrading pciids from 20260430 to 20260514...
[7/48] Extracting pciids-20260514: ..... done
[8/48] Upgrading php83 from 8.3.30 to 8.3.31...
[8/48] Extracting php83-8.3.31: .......... done
[9/48] Upgrading php83-mbstring from 8.3.30 to 8.3.31...
[9/48] Extracting php83-mbstring-8.3.31: .......... done
[10/48] Deinstalling opnsense-26.1.8_5...
Stopping configd...done
Resetting root shell
Updating /etc/shells
Unhooking from /etc/rc
Unhooking from /etc/rc.shutdown
[10/48] Deleting files for opnsense-26.1.8_5: .......... done
[11/48] Upgrading ca_root_nss from 3.117_2 to 3.124...
[11/48] Extracting ca_root_nss-3.124: ..... done
[12/48] Upgrading openvpn from 2.6.20 to 2.7.4...
===> Creating groups
Using existing group 'openvpn'
===> Creating users
Using existing user 'openvpn'
[12/48] Extracting openvpn-2.7.4: .......... done
[13/48] Upgrading opnsense-update from 26.1.7_1 to 26.1.9...
[13/48] Extracting opnsense-update-26.1.9: .......... done
[14/48] Upgrading php83-ctype from 8.3.30 to 8.3.31...
[14/48] Extracting php83-ctype-8.3.31: ........ done
[15/48] Upgrading php83-curl from 8.3.30 to 8.3.31...
[15/48] Extracting php83-curl-8.3.31: .......... done
[16/48] Upgrading php83-dom from 8.3.30 to 8.3.31...
[16/48] Extracting php83-dom-8.3.31: .......... done
[17/48] Upgrading php83-filter from 8.3.30 to 8.3.31...
[17/48] Extracting php83-filter-8.3.31: ......... done
[18/48] Upgrading php83-gettext from 8.3.30 to 8.3.31...
[18/48] Extracting php83-gettext-8.3.31: ........ done
[19/48] Upgrading php83-ldap from 8.3.30 to 8.3.31...
[19/48] Extracting php83-ldap-8.3.31: ........ done
[20/48] Upgrading php83-pcntl from 8.3.30 to 8.3.31...
[20/48] Extracting php83-pcntl-8.3.31: ......... done
[21/48] Upgrading php83-pdo from 8.3.30 to 8.3.31...
[21/48] Extracting php83-pdo-8.3.31: .......... done
[22/48] Upgrading php83-session from 8.3.30 to 8.3.31...
[22/48] Extracting php83-session-8.3.31: .......... done
[23/48] Upgrading php83-simplexml from 8.3.30 to 8.3.31...
[23/48] Extracting php83-simplexml-8.3.31: ......... done
[24/48] Upgrading php83-sockets from 8.3.30 to 8.3.31...
[24/48] Extracting php83-sockets-8.3.31: .......... done
[25/48] Upgrading php83-sqlite3 from 8.3.30 to 8.3.31...
[25/48] Extracting php83-sqlite3-8.3.31: ......... done
[26/48] Upgrading php83-xml from 8.3.30 to 8.3.31...
[26/48] Extracting php83-xml-8.3.31: ......... done
[27/48] Upgrading php83-zlib from 8.3.30 to 8.3.31...
[27/48] Extracting php83-zlib-8.3.31: ........ done
[28/48] Deinstalling py313-duckdb-1.5.2...
[28/48] Deleting files for py313-duckdb-1.5.2: .......... done
[29/48] Deinstalling py313-pandas-2.3.3_1,1...
[29/48] Deleting files for py313-pandas-2.3.3_1,1: .......... done
[30/48] Deinstalling py313-bottleneck-1.6.0_1...
[30/48] Deleting files for py313-bottleneck-1.6.0_1: .......... done
[31/48] Deinstalling py313-numexpr-2.14.1_1...
[31/48] Deleting files for py313-numexpr-2.14.1_1: .......... done
[32/48] Deinstalling py313-numpy1-1.26.4_3...
[32/48] Deleting files for py313-numpy1-1.26.4_3: .......... done
[33/48] Upgrading python313 from 3.13.13 to 3.13.13_1...
[33/48] Extracting python313-3.13.13_1: .......... done
[34/48] Upgrading py313-idna from 3.13 to 3.15...
[34/48] Extracting py313-idna-3.15: .......... done
[35/48] Installing py313-numpy-2.4.4_3,1...
[35/48] Extracting py313-numpy-2.4.4_3,1: .......... done
[36/48] Installing py313-bottleneck-1.6.0_2...
[36/48] Extracting py313-bottleneck-1.6.0_2: .......... done
[37/48] Installing py313-numexpr-2.14.1_2...
[37/48] Extracting py313-numexpr-2.14.1_2: .......... done
[38/48] Upgrading py313-pytz from 2025.2_1,1 to 2026.2,1...
[38/48] Extracting py313-pytz-2026.2,1: .......... done
[39/48] Installing py313-pandas-2.3.3_2,1...
[39/48] Extracting py313-pandas-2.3.3_2,1: .......... done
[40/48] Installing py313-duckdb-1.5.2_1...
[40/48] Extracting py313-duckdb-1.5.2_1: .......... done
[41/48] Upgrading py313-ujson from 5.12.0 to 5.12.1...
[41/48] Extracting py313-ujson-5.12.1: .......... done
[42/48] Upgrading py313-urllib3 from 2.6.3,1 to 2.7.0,1...
[42/48] Extracting py313-urllib3-2.7.0,1: .......... done
[43/48] Upgrading py313-botocore from 1.43.2 to 1.43.7...
[43/48] Extracting py313-botocore-1.43.7: .......... done
[44/48] Upgrading py313-boto3 from 1.43.2 to 1.43.7...
[44/48] Extracting py313-boto3-1.43.7: .......... done
[45/48] Upgrading py313-requests from 2.33.1 to 2.34.2...
[45/48] Extracting py313-requests-2.34.2: .......... done
[46/48] Upgrading suricata from 8.0.4 to 8.0.5...
[46/48] Extracting suricata-8.0.5: .........
pkg-static: Fail to create /var/log/suricata:No space left on device
[46/48] Extracting suricata-8.0.5... done
Flushing temporary package files... done
eval: /usr/local/etc/rc.restart_webgui: not found
***REBOOT***
```
Looks like it removed core on both, not sure why. Ignore the suricata stuff, not entirely set up at the moment. Both machines are identical.

**Comment by doveman** [comment]
Similar thing happened to me this week, although it didn't happen when I updated. I haven't updated it in a couple of weeks at least but I came home last night and found the Web UI wasn't working and I got the "/usr/local/libexec/opnsense-auth: not found" error when trying to login via the console. There were also other errors about Kea (failed to initialise Kea server; failed to select interface: interface 'vlan0.4.21' doesn't exist in the system)  and monit and hostwatch.

I'm running it in a VM under Proxmox, so I tried restoring from my backups for the last four days but they all had the same problem, so I cloned it to a new VM and reinstalled from the 26.1.6 ISO that I had, telling it to import my config from the existing disk before wiping it and doing the install and that worked. I just had to reinstall my plugins after and updated to 26.1.10 no problem.
